### PR TITLE
check fake/real mismatches during real tensor prop

### PR DESCRIPTION
**Pull request by pianpwk**
Summary:
While testing exportability for PT2 Inference models, we found various cases of invalid op inputs during tracing, for example errors like: `a and b must have same reduction dim`, `expected scalar type Long but found Int`, etc. Looking more closely, these happened to due the same few meta kernels & eager kernels producing mismatched outputs upstream (e.g. different output tensor dtype, int output).

Adding checks to catch mismatched outputs in real tensor prop upstream, so errors are raised at the mismatched op, instead of the downstream ops taking them as inputs. Relies a lot on utils from [CrossRefFakeMode](https://github.com/pytorch/pytorch/blob/929797dedbf23376123ce95230c01a7e3b71e130/torch/_subclasses/fake_utils.py#L78)

Follow ups: could add more checks, and maybe have a flag to only enable these for cases like draft mode, so perf doesn't suffer?

Test Plan: test_export, test_fake_tensor

Differential Revision: D64210055

cc @ezyang @penguinwu @bobrenjc93

### CODE REVIEW

**Comment by ezyang** [comment]
You're gonna submit fixes for the real/fake mismatches right? lol

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -1918,6 +1918,22 @@ def go(t: object, real_t: Tensor) -> None:
...
+            # use real values to check for potentially mismatched fake/real kernels
+            def _fake_real_mismatch(fake, real) -> Optional[bool]:
+                if isinstance(fake, (SymInt, SymFloat)):
+                    # symbolic expression, ask ShapeEnv if it's sure of inequality.
+                    assert self.shape_env is not None
+                    return (
+                        self.shape_env._maybe_evaluate_static(
+                            sympy.Ne(fake.node.expr, real)
+                        )
+                    ) is sympy.S.true
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:1930]: I think you can do this better. Specifically, in real tensor propagation mode, I should always know the underlying value of all unbacked SymInts. So I should *always* be able to do an *exact* match between the real size, and the symbolic expression /after/ I've done all the substitutions (this is akin to hint, but you also want to substitute in all of the unbacked vars too, I don't think we have a direct method for this but it's easy enough to add).

It's a little more tricky if you want to handle cases where I'm missing a binding for an unbacked symint for some reason, but I'd probably just skip the test in that case.

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -1934,6 +1950,24 @@ def go(t: object, real_t: Tensor) -> None:
                 else:
                     tree_map_(go, fake_out, real_out)
 
+                if isinstance(fake_out, torch.Tensor):
+                    assert fake_out.dtype == real_out.dtype, (
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:1954]: There are a lot more metadata properties than this?

**Comment by ezyang** [comment]
I'm positive about the change but it needs some improvements.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -2121,6 +2121,8 @@ def _compute_reduction_shape(self, dims, keepdim):
 def device_hint(tensor) -> "str":
     if isinstance(tensor, torch._subclasses.FakeTensor):
         return tensor.fake_device.type
+    elif hasattr(tensor, "device") and hasattr(tensor.device, "type"):
+        return tensor.device.type
```
Comment (ezyang) [inline torch/_meta_registrations.py:2125]: this feels potentially bc breaking since meta tensors will now report meta rather than cuda

**Comment by pianpwk** [comment]
> You're gonna submit fixes for the real/fake mismatches right? lol

Yepp, though they're internal custom ops, I've put up a diff

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (pianpwk) [reply torch/_subclasses/fake_tensor.py]: Just updated, to follow what CrossRefFakeMode does

File: test/test_fake_tensor.py [hunk 4]
```diff
@@ -763,6 +763,8 @@ def run_meta():
         with patch.object(torch._functorch.config, "fake_tensor_allow_meta", False):
             self.assertRaises(Exception, run_meta)
 
+    # FakeTensor device vs. cuda implementations have mismatched size in 2nd output of _embedding_bag op
+    @expectedFailurePropagateRealTensors
```
Comment (pianpwk) [inline test/test_fake_tensor.py:767]: We're running this test with cpu FakeTensors, but for real tensors the device defaults to "cuda", and the offsets2bag output (which is actually not returned in the `torch.nn.EmbeddingBag.forward`, just the intermediate `aten._embedding_bag.default` op), has a different size b/w the cpu & cuda implementations:

https://github.com/pytorch/pytorch/blob/9f81270d7589fd7fa98dc247ae4b1b7ab239ca3c/torch/_meta_registrations.py#L3423-L3435

Comment (ezyang) [reply test/test_fake_tensor.py:767]: yeah, you can't just default it. need to actually have the correct devices

File: test/export/test_export.py [hunk 5]
```diff
@@ -938,6 +938,137 @@ def forward(self, x):
...
+
+            class M(torch.nn.Module):
+                def forward(self, a, b):
+                    return torch.ops.mylib.foo(a, b)
+
+            @torch.library.impl("mylib::foo", "cpu", lib=lib)
+            def foo_impl(a, b):
+                return a + b
+
+            @torch.library.register_fake("mylib::foo")
```
Comment (zou3519) [inline test/export/test_export.py:958]: nit: lib=lib

File: test/export/test_export.py [hunk 6]
```diff
@@ -938,6 +938,137 @@ def forward(self, x):
...
+            # test alias case
+            class M(torch.nn.Module):
+                def forward(self, a):
+                    return torch.ops.mylib.foo_alias(a)
+
+            @torch.library.impl("mylib::foo_alias", "cpu", lib=lib)
+            def foo_impl(a):
+                return a * 2
+
+            @torch.library.register_fake("mylib::foo_alias")
```
Comment (zou3519) [inline test/export/test_export.py:1038]: lib=lib (unless it doesn't work?)

File: torch/_subclasses/fake_tensor.py [hunk 7]
```diff
@@ -2045,6 +2064,67 @@ def go(t: object, real_t: Tensor) -> None:
...
+                # check aliasing info
+                try:
+                    _check_alias_info(
+                        "Real tensor propagation found an aliasing",
+                        (flat_real_out := pytree.tree_leaves(real_out)),
+                        (real_args, real_kwargs),
+                        (flat_fake_out := pytree.tree_leaves(fake_out)),
+                        (args, kwargs),
+                    )
+                except AssertionError as exc:
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2076]: You probably want your own exception type here that you can catch and raise. Otherwise, it'll be difficult to tell the difference between an "aliasing mismatch" vs a "development assert"

File: torch/_subclasses/fake_tensor.py [hunk 8]
```diff
@@ -2045,6 +2064,67 @@ def go(t: object, real_t: Tensor) -> None:
...
+                    raise AssertionError(
+                        "Real tensor propagation found an aliasing mismatch between fake output "
+                        f"and real output for func: {func}"
+                    ) from exc
+
+                for i, (_real_out, _fake_out) in enumerate(
+                    zip(flat_real_out, flat_fake_out)
+                ):
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2084]: Why can't we refactor all of https://github.com/pytorch/pytorch/blob/929797dedbf23376123ce95230c01a7e3b71e130/torch/_subclasses/fake_utils.py#L133-L196 into a single function `check_fake_real(fakes, reals)` and call it here? 

If the problem is that that function doesn't have everything we want to check, then we should improve it to check everything that we want to check

Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:2084]: For example, fake_utils.py has a storage offset check that I don't see here

Comment (pianpwk) [reply torch/_subclasses/fake_tensor.py]: Ah just added. The checks in fake_utils.py were exposing some larger mismatches (bugs?), e.g. strides & requires_grad, so I optionally turned them off for now.

I also added an output value check for non-tensor values - not sure if we should have that in CrossRefFakeMode too?

Lastly it seems like the sizes check in fake_utils.py differs a bit (accomodates for broadcasting, isn't as strict on equality), so I left that alone.

File: test/export/test_export.py [hunk 9]
```diff
@@ -995,6 +995,141 @@ def forward(self, x):
...
+                    return torch.ops.mylib.foo(a, b)
+
+            @torch.library.impl("mylib::foo", "cpu", lib=lib)
+            def foo_impl(a, b):
+                return a + b
+
+            @torch.library.register_fake("mylib::foo", lib=lib)
+            def foo_fake_impl(a, b):
+                m, n = a.shape
+                return torch.empty(n, m)  # incorrectly permute
```
Comment (zou3519) [inline test/export/test_export.py:1020]: nit: use torch.library.custom_op unless you're explicitly testing the low-level torch.library APIs

File: torch/_meta_registrations.py (hunk 3 above)
Comment (zou3519) [reply torch/_meta_registrations.py:2139]: @pianpwk Why do we need this? Also did we resolve the bc breaking concern?

File: torch/_subclasses/fake_tensor.py [hunk 10]
```diff
@@ -2035,6 +2045,26 @@ def go(t: object, real_t: Tensor) -> None:
...
+                    if (
+                        not fake.node.expr.free_symbols
+                        - self.shape_env.var_to_val.keys()
+                        - self.shape_env.unbacked_var_to_val.keys()
+                    ):
+                        assert (
+                            self.shape_env._maybe_evaluate_static(
+                                sympy.Eq(fake.node.expr, real), compute_hint=True
+                            )
+                        ) is sympy.S.true
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2062]: Gonna need @ezyang to check if this is the right sequence of operations

File: torch/_subclasses/fake_tensor.py [hunk 11]
```diff
@@ -2035,6 +2045,26 @@ def go(t: object, real_t: Tensor) -> None:
...
+                elif isinstance(
+                    fake, (int, float, bool)
+                ):  # concrete value, check direct equality
+                    assert fake == real
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2066]: nit: assert False if we get out of this branch?

File: torch/_subclasses/fake_tensor.py (hunk 7 above)
Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:2093]: Sorry my suggestion was that `_check_alias_info` should return a special kind of exception that we then catch. Like the following:
```py
try:
    _check_alias_info(...)
except FakeCrossRefError as exc:
    ...
```
This way there is no confusing `assert` statements in `_check_alias_info` from "your alias info checks failed" errors

File: torch/_subclasses/fake_tensor.py [hunk 12]
```diff
@@ -2051,6 +2081,65 @@ def go(t: object, real_t: Tensor) -> None:
...
+                try:
+                    _check_alias_info(
+                        "Real tensor propagation found",
+                        real_out,
+                        (real_args, real_kwargs),
+                        fake_out,
+                        (args, kwargs),
+                    )
+                except AssertionError as exc:
+                    raise FakeRealKernelMismatchException(
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2094]: Why did we need a new exception type here?

File: torch/_subclasses/fake_tensor.py [hunk 13]
```diff
@@ -2051,6 +2081,65 @@ def go(t: object, real_t: Tensor) -> None:
...
+                            _check_fake_real_tensors(
+                                _fake_out,
+                                _real_out,
+                                context="Real tensor propagation found",
+                                sizes=False,  # manual check below
+                                strides=False,  # skip strides
+                                storage_offset=True,
+                                requires_grad=False,  # issues with FakeTensorConverter preserving requires_grad
+                            )
+                        except (AssertionError, RuntimeError) as exc:
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2115]: Ditto here (and for the other try-catch), we should have something like:
```
try: _check_fake_real_tensors(...) except FakeCrossRefError as exc: ...
```

**Review by zou3519 — COMMENTED** [review]
Let's raise special exceptions from the `_check_*` functions so that they can be caught (instead of caching AssertionErrors and hoping they are the exceptions we want to surface to the user). Other than that, this looks good to me (left some questions)

File: torch/_meta_registrations.py (hunk 3 above)
Comment (pianpwk) [reply torch/_meta_registrations.py:2139]: I think what it was is we were running cross-checks with cpu FakeTensors and real tensors, and `device_hint()` would return "cpu" on FakeTensors, but "cuda" for real tensors, and ran into meta kernels that returned different output formats for cpu/cuda.

This is now not BC-breaking (but a bit hacky), because meta tensors now return "cuda" regardless; I'm not sure what stuff, but some stuff relies on this that I don't want to break.

File: torch/_subclasses/fake_tensor.py (hunk 12 above)
Comment (pianpwk) [reply torch/_subclasses/fake_tensor.py:2094]: I guess we don't really.. I think I just misunderstood your comment above (on aliasing)

File: torch/_prims_common/__init__.py [hunk 14]
```diff
@@ -168,27 +173,27 @@ def compare_tensor_meta(
...
 
         if a.storage_offset() != b.storage_offset():
             msg = f"Storage offset mismatch! Storage offsets are {a.storage_offset()} and {b.storage_offset()}!"
-            raise RuntimeError(msg)
+            raise TensorCrossRefException(msg)
 
     if check_conj:
         if a.is_conj() != b.is_conj():
-            raise RuntimeError(
+            raise TensorCrossRefException(
```
Comment (zou3519) [inline torch/_prims_common/__init__.py:191]: nit: call it FakeCrossRef?

File: torch/_subclasses/fake_tensor.py [hunk 15]
```diff
@@ -2063,6 +2093,65 @@ def go(t: object, real_t: Tensor) -> None:
...
+                                f"fake tensor {_fake_out} and real tensor {_real_out}, "
+                                f" at output index {i}, for func: {func}"
+                            ) from exc
+
+                        for j, (s_fake, s_real) in enumerate(
+                            zip(_fake_out.size(), _real_out.size())
+                        ):
+                            try:
+                                _check_fake_real_vals(s_fake, s_real)
+                            except AssertionError as exc:
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2139]: Can this be
```
try:
    _check_fake_real_vals(s_fake, s_real)
except TensorCrossRefException as exc:
```
too? I suggest renaming TensorCrossRefException to FakeCrossRefError and then using "FakeCrossRefError" to indicate that there's a discrepancy between real vs fake

File: torch/_subclasses/fake_tensor.py [hunk 16]
```diff
@@ -2063,6 +2093,65 @@ def go(t: object, real_t: Tensor) -> None:
...
+                                raise TensorCrossRefException(
+                                    f"Real tensor propagation found an output size mismatch between "
+                                    f"fake shape {s_fake} and real shape {s_real}, at output "
+                                    f"index {i}, dimension {j} for func: {func}"
+                                ) from exc
+                    else:
+                        try:
+                            _check_fake_real_vals(_fake_out, _real_out)
+                        except AssertionError as exc:
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2148]: Ditto here

File: torch/_prims_common/__init__.py (hunk 14 above)
Comment (pianpwk) [reply torch/_prims_common/__init__.py:191]: is it ok to keep it like this? I'm raising this in `_prims_common.py/compare_tensor_meta()` too, which isn't necessarily for FakeTensors

Comment (zou3519) [reply torch/_prims_common/__init__.py:191]: Ah I didn't realize. TensorMetadataMismatchError maybe?

Comment (zou3519) [reply torch/_prims_common/__init__.py:191]: Or maybe just MetadataMismatchError if we didn't want Tensor in the name

Comment (pianpwk) [reply torch/_prims_common/__init__.py]: thanks, just changed!

File: torch/_subclasses/fake_tensor.py [hunk 17]
```diff
@@ -2047,6 +2057,30 @@ def go(t: object, real_t: Tensor) -> None:
...
+                        not fake.node.expr.free_symbols
+                        - self.shape_env.var_to_val.keys()
+                        - self.shape_env.unbacked_var_to_val.keys()
+                    ):
+                        if self.shape_env._maybe_evaluate_static(
+                            sympy.Eq(fake.node.expr, real), compute_hint=True
+                        ) is not sympy.S.true:
+                            raise MetadataMismatchError(
+                                f"mismatch between fake value {fake} and real value {real} "
+                            )
```
Comment (zou3519) [inline torch/_subclasses/fake_tensor.py:2075]: Might want Ed to check this, I'm not sure what this is doing

**Review by zou3519 — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]
